### PR TITLE
374 current time settings not working for long duration 600 years

### PR DESCRIPTION
**Pull request by OnnoEbbens**

### CODE REVIEW

**Comment by OnnoEbbens** [comment]
failing test are due to a bug in hydropandas that will be solved by: https://github.com/ArtesiaWater/hydropandas/pull/246

**Comment by rubencalje** [comment]
Tests now fail because of numpy version. I solved this in PR #373 by pinning the numpy version for ci in pyproject.toml.

File: nlmod/dims/time.py [hunk 1]
```diff
@@ -244,9 +307,138 @@ def set_ds_time(
...
+    nstp : int or array-like, optional
+        number of steps per stress period, stored in ds.attrs, default is 1
+    tsmult : float, optional
+        timestep multiplier within stress periods, stored in ds.attrs, default is 1.0
+
+    Returns
+    -------
+    _type_
+        _description_
+    """
```
Comment (dbrakenhoff) [inline nlmod/dims/time.py:441]: add return info?

File: nlmod/dims/time.py [hunk 2]
```diff
@@ -518,6 +719,9 @@ def dataframe_to_flopy_timeseries(
     append=False,
 ):
     assert not df.isna().any(axis=None)
+    assert (
+        ds.time.dtype.kind == "M"
```
Comment (dbrakenhoff) [inline nlmod/dims/time.py:723]: What does this M mean?

Comment (dbrakenhoff) [reply nlmod/dims/time.py:723]: https://numpy.org/doc/2.0/reference/generated/numpy.dtype.kind.html

File: nlmod/dims/time.py [hunk 3]
```diff
@@ -244,9 +307,138 @@ def set_ds_time(
...
 
+    # add steady, nstp and tsmult to dataset
+    ds = set_time_variables(
+        ds, start, time, steady, steady_start, time_units, nstp, tsmult
+    )
+
+    return ds
+
+
+def set_ds_time_numerical(
```
Comment (dbrakenhoff) [inline nlmod/dims/time.py:322]: very nitpicky, but set_ds_time_numeric() might be grammatically slightly better?

File: nlmod/dims/time.py [hunk 4]
```diff
@@ -244,9 +307,138 @@ def set_ds_time(
...
+    time_units : str, optional
+        time units, by default "DAYS"
+    nstp : int or array-like, optional
+        number of steps per stress period, stored in ds.attrs, default is 1
+    tsmult : float, optional
+        timestep multiplier within stress periods, stored in ds.attrs, default is 1.0
+
+    Returns
+    -------
+    _type_
```
Comment (dbrakenhoff) [inline nlmod/dims/time.py:439]: add return info

File: nlmod/dims/time.py [hunk 5]
```diff
@@ -560,6 +764,12 @@ def ds_time_to_pandas_index(ds, include_start=True):
         pandas datetime index
     """
     if include_start:
-        return ds.time.to_index().insert(0, pd.Timestamp(ds.time.start))
+        if ds.time.dtype.kind == "M":
```
Comment (dbrakenhoff) [inline nlmod/dims/time.py:767]: maybe add comment what M means:  datetime-type

File: tests/test_016_time.py [hunk 6]
```diff
@@ -32,3 +39,153 @@ def test_ds_time_from_tdis_settings():
 
 def test_get_time_step_length():
     assert (nlmod.time.get_time_step_length(100, 2, 1.5) == np.array([40, 60])).all()
+
+
+def test_time_options():
```
Comment (dbrakenhoff) [inline tests/test_016_time.py:44]: nice, very thorough :)!

File: nlmod/sim/sim.py [hunk 7]
```diff
@@ -89,12 +90,23 @@ def get_tdis_perioddata(ds, nstp="nstp", tsmult="tsmult"):
...
+            perlen.extend(np.diff(ds["time"]) / deltat)
+    elif ds.time.dtype.kind == "O":
+        perlen = [
+            (ds["time"].data[0] - _pd_timestamp_to_cftime(pd.Timestamp(ds.time.start)))
+            / deltat
+        ]
+        if len(ds["time"]) > 1:
+            perlen.extend(np.diff(ds["time"]) / deltat)
+    elif ds.time.dtype.kind in ["i", "f"]:
+        perlen = ds["time"].values
```
Comment (dbrakenhoff) [inline nlmod/sim/sim.py:109]: Is this right? Shouldn't there be a diff here as well?

File: nlmod/gwf/surface_water.py [hunk 8]
```diff
@@ -1111,6 +1111,8 @@ def add_season_timeseries(
     summer_name : str, optional
         The name of the time-series with ones in summer. The default is "summer".
     """
```
Comment (dbrakenhoff) [inline nlmod/gwf/surface_water.py:1113]: opposite check? kind == "M"?

**Review by dbrakenhoff — COMMENTED** [review]
Really nice work! I left some comments here and there. 

One more general comment, that might be a bit controversial :see_no_evil: as you put in quite some work to get cftime to work automatically:

There seems to be some repetition of cftime conversion in each of the different paths to construct a time index. This functionality is nice, but also adds quite some complexity to `set_ds_time()`. Just wondering whether you considered trapping the pandas OutofBounds error, and just returning an error message to the user to convert the input (start or time) to cftime types? And then having just the one path for constructing a time index using cftime? 

Just a thought, so feel free to dismiss it since the automatic functionality is already fully operational now :).

**Comment by OnnoEbbens** [comment]
Thanks for the review and you are right. I was trapped by the sunk cost fallacy . I will update the PR

**Pull request merged**